### PR TITLE
Allow features to be cyclical

### LINKED ISSUE #3796: Cargo doesn't allow cyclic features
If a crate has two features `foo` and `bar` which depend on each other, Cargo will error.

```toml
[features]
foo = ["bar"]
bar = ["foo"]
```
```
error: Cyclic feature dependency: feature `foo` depends on itself
```
However, I ended up needing this in `winapi`. In order to allow people to compile only the bits they need to save time, `winapi` has a feature for each header. However since headers depend on each other, the features in turn had to depend on each other to ensure that `winapi` would compile correctly. Unfortunately sometimes those headers would have cyclic dependencies on each other, causing Cargo to complain. There were several workarounds I tried such as having a feature that is common to a set of headers which depend on each other, but I eventually settled on doing feature dependency resolution myself via a build script. https://github.com/retep998/winapi-rs/blob/dev/build.rs#L175-L191

### PR DESCRIPTION
**Pull request by wesleywiser**
Fixes #3796

### CODE REVIEW

File: src/cargo/core/resolver/mod.rs [hunk 1]
```diff
@@ -907,15 +907,22 @@ fn build_features<'a>(s: &'a Summary, method: &'a Method)
...
+
                 used.insert(feat);
                 match s.features().get(feat) {
                     Some(recursive) => {
                         // This is a feature, add it recursively.
                         for f in recursive {
+                            if f == feat {
+                                bail!("Cyclic feature dependency: feature `{}` depends \
+                                        on itself", feat);
+                            }
```
Comment (matklad) [inline src/cargo/core/resolver/mod.rs]: This forbids direct dependency of feature on itself, right? Let's add a test case for this then?

File: src/cargo/core/resolver/mod.rs [hunk 2]
```diff
@@ -907,15 +907,22 @@ fn build_features<'a>(s: &'a Summary, method: &'a Method)
             }
             None => {
                 let feat = feat_or_package;
+
+                //if this feature has already been added, then just return Ok
                 if !visited.insert(feat) {
```
Comment (matklad) [inline src/cargo/core/resolver/mod.rs]: There's `visited.remove(feat)` down the line, and it seems suspicious to me.

Can it lead to visiting and walking the same feature more then once? Does anything break if we just remove the `.remove`? I would think it was needed for better error reporting only.

Comment (wesleywiser) [reply src/cargo/core/resolver/mod.rs]: I will look into that later today.

File: src/cargo/core/resolver/mod.rs (hunk 1 above)
Comment (wesleywiser) [reply src/cargo/core/resolver/mod.rs]: Yes that's correct. There was an existing test for this [`cyclic_feature()`](https://github.com/rust-lang/cargo/pull/4473/files#diff-40231dbe3cfb4cb45dee09f272e30cb6L451) so I left it as-is.

File: src/cargo/core/resolver/mod.rs (hunk 2 above)
Comment (wesleywiser) [reply src/cargo/core/resolver/mod.rs]: I agree. It makes sense to remove it. I've pushed a fix for that.

**Comment by wesleywiser** [comment]
Pinging @matklad in case they didn't see the updated commit.

**Comment by matklad** [comment]
Thanks for the poke @wesleywiser !

@bors r+

**Comment by wesleywiser** [comment]
No problem @matklad. Thanks for the review!

**Pull request merged**